### ISSUE TITLE
`cilium bpf ct list global -o json` prints map details, not ct entries

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#2976]
```
$ sudo cilium bpf ct list global | head -n 5 
TCP IN [f00d::a0f:0:0:4870]:80 -> [f00d::a0f:0:0:4264]:34274  expires=71101 rx_packets=6 rx_bytes=615 tx_packets=4 tx_bytes=581 flags=0 revnat=0 proxyport=0 src_sec_id=42212
TCP OUT [f00d::a0f:0:0:ad]:80 -> [f00d::a0f:0:0:dc06]:34770  expires=70766 rx_packets=0 rx_bytes=0 tx_packets=3 tx_bytes=282 flags=0 revnat=0 proxyport=0 src_sec_id=42212
ICMPv6 OUT [f00d::a0f:0:0:4870]:32768 -> [f00d::a0f:0:0:1150]:0  expires=71092 rx_packets=10 rx_bytes=1180 tx_packets=10 tx_bytes=1180 flags=0 revnat=0 proxyport=0 src_sec_id=41924
TCP OUT [f00d::a0f:0:0:ecfe]:80 -> [f00d::a0f:0:0:f8a]:48700  expires=70964 rx_packets=4 rx_bytes=581 tx_packets=6 tx_bytes=615 flags=0 revnat=0 proxyport=0 src_sec_id=41924
TCP OUT [f00d::a0f:0:0:ecfe]:80 -> [f00d::a0f:0:0:10da]:57812  expires=70921 rx_packets=0 rx_bytes=0 tx_packets=3 tx_bytes=282 flags=0 revnat=0 proxyport=0 src_sec_id=42212
```

```
$ sudo cilium bpf ct list global -o json       
{
  "MapType": 1,
  "KeySize": 40,
  "ValueSize": 48,
  "MaxEntries": 1000000,
  "Flags": 0,
  "OwnerProgType": 0,
  "NonPersistent": false
}
{
  "MapType": 1,
  "KeySize": 14,
  "ValueSize": 48,
  "MaxEntries": 1000000,
  "Flags": 0,
  "OwnerProgType": 0,
  "NonPersistent": false
}
```

This command is intended to print the contents of the map, not the details of the IPv4/IPv6 map structures.

**Comment by jmuzsik** [comment]
Hi, I'd be willing to do this but I'd need much more words and/or resources to understand where to find the problem/how to solve it

**Comment by joestringer** [comment]
@jMuzsik awesome! Let me see how I can help with this.

In the output in the report above, the first command's output prints a bunch of entries that have details such as IP addresses and so on. They're entries in the map. However, when using `-o json` it's not printing the entries in the map, but instead a JSON formatting of the details of the map itself.

The issue here is that the variable `m` in the code below is of type `*bpf.Map`, so when it's passed into the JSON formatter it prints the json-formatted structure of the `bpf.Map` rather than iterating the map and getting its contents. The `bpf.Map` type has a `Dump()` function which can turn the map's contents into a native golang map (of type `map[string][]string`). If that is passed into the `PrintOutput()` function then the output should print the actual elements in the map.

https://github.com/cilium/cilium/blob/8104d12384805ad576af194de5f98e0c5dea5794/cilium/cmd/bpf_ct_list.go#L64

To be able to test it out and make sure it's printing values that make sense, I'd recommend taking a look at the [getting started guides](http://docs.cilium.io/en/v1.0.0-rc4/gettingstarted/), and pick one (say, the minikube or docker compose one), then once you get to the point where you have enforced some L3 (or L3/L4) policy, run the commands above and see what the output looks like. The JSON should print a bunch of details like IP addresses, identities, expiry times, packet counts etc.

**Comment by jmuzsik** [comment]
Having errors in getting cilium set up on my machine and am trouble shooting at the moment. Once finished if I run into trouble in implementing the issue I'll check back on here or on the Cilium slack, thanks!

**Comment by tgraf** [comment]
@jMuzsik Are you using the vagrant box? (http://docs.cilium.io/en/v1.0.0-rc4/contributing/#setting-up-a-development-environment) It's by far the easiest way to get a dev environment up and running.

**Comment by jmuzsik** [comment]
Yes I am but I thought it was because I had issues related to a previous installation of the virtualbox but rather I believe that it may be a problem in the getting-started that needs to be updated:

Its a simple bug, and I'll post it later tonight as an actual issue.

This is the gist of it:

In the `Vagrantfile` the Vagrant config uses  ```bento/ubuntu-17.04``` but this causes an error b/c this version of ubuntu will fail during the download of docker compose - check out [this] from ubuntu forums (https://ubuntuforums.org/showthread.php?t=2382832). And an example from my terminal:

<img width="1280" alt="screen shot 2018-03-05 at 17 33 59" src="https://user-images.githubusercontent.com/29470576/37010928-db1b58fe-20bb-11e8-97f8-c751382243b4.png">

So basically all instances in the examples of ```bento/ubuntu-17.04``` must be replaced with ```bento/ubuntu-17.10```

It's actually incredibly simple, I only need to replace it in two instances: in the `Vagrantfile` of `examples/mesos `and `examples/getting-started`.

So ought I simply do this while doing this issue or make it a separate issue?

**Comment by ianvernon** [comment]
@jMuzsik - the issue with the VM you are running is fixed by https://github.com/cilium/cilium/pull/3020 . However, the VM you are running is not the best VM for compiling / testing Cilium. Please refer to the [using the provided Vagrantfile](http://docs.cilium.io/en/v1.0.0-rc4/contributing/#using-the-provided-vagrantfile) section.

**Comment by jmuzsik** [comment]
Hi, sorry but this is impossible for me, it is not as simple as I supposed. 

First of all, something like this has to be written within `cilium/cmd/bpf_ct_list.go`:

```go
	bpfCtList := make(map[string][]string)
	if err := ctMap.CtMap.Dump(bpfCtList); err != nil {
		os.Exit(1)
	}
```

As, this form of function always happens in other instances of outputting JSON through the command line. Such as in:

`bpf_endpoint_list.go` or `bpf_tunnel_list.go`

Then,  the best reference I saw that had a similarity to what needed to be done was within `bpf_endpoint_list.go`'s function: ` lxcmap.LXCMap.Dump(bpfEndpointList)`

This first creates the specified bpf map then Dumps into the matrix.

As so:

```go
var (
	// LXCMap represents the BPF map for endpoints
	LXCMap = bpf.NewMap(MapName,
		bpf.MapTypeHash,
		int(unsafe.Sizeof(EndpointKey{})),
		int(unsafe.Sizeof(EndpointInfo{})),
		MaxKeys,
		0,
		func(key []byte, value []byte) (bpf.MapKey, bpf.MapValue, error) {
			k, v := EndpointKey{}, EndpointInfo{}

			if err := bpf.ConvertKeyValue(key, value, &k, &v); err != nil {
				return nil, nil, err
			}

			return k, v, nil
		},
	)
)
```

A function like this does not exist in `ctmap.go` which I believe must be created. From my experiments (which I am surely incapable of doing in the precise way you guys need) it appears that not only that function, but other functions to fit the type and the requirements of `ctMap` must be created. 

Also, I am quite sure that within `pkg/bpf` a new file needs to be created to handle the creation of the CtMap.

Sorry and thanks for letting me put my hand into trying to implement this, I learned quite a bit in the process, and if there are any simple problems I'd be happy to give it a try. Thanks!

**Comment by joestringer** [comment]
@jMuzsik thanks for digging in to take a look, it seems that my instructions were underestimating the work required as I didn't realise how different the approach is in CT map vs. other map types. Regardless, thanks for trying it out, I think that your analysis should be useful in future for implementing this functionality. I hope that you were able to learn some things along the way.

One minor comment, there should not need to be any new files in `pkg/bpf`, the `pkg/maps/ctmap` should be able to be adjusted to provide access to golang structs representing the values which can then be serialised into strings, but exactly how to do this without significantly reworking the map type seems a bit tricky. Note that for ctmap, the name is not static - it depends on the `CONNTRACK_LOCAL` config option, so we can't just define a variable for the map once and reuse that everywhere.

**Comment by joestringer** [comment]
FYI https://github.com/cilium/cilium/pull/5480 is refactoring the CT map code to make the CT maps more accessible, so this bug should become much easier to address once that goes in.

**Comment by jmuzsik** [comment]
Will look into it!
